src/cgutils.cpp

```diff
@@ -2504,21 +2504,45 @@ static jl_cgval_t emit_unionload(jl_codectx_t &ctx, Value *addr, Value *ptindex,
     return mark_julia_slot(fsz > 0 ? addr : nullptr, jfty, tindex, tbaa);
 }
 
+static const char *get_fieldname(jl_datatype_t *jt, unsigned idx)
+{
+    if (jl_is_namedtuple_type(jt)) {
+        auto names = jl_tparam0(jt);
+        assert(jl_is_tuple(names));
+        if (idx < jl_nfields(names)) {
+            auto name = jl_fieldref(names, idx);
+            assert(jl_is_symbol(name));
+            return jl_symbol_name((jl_sym_t*)name);
+        }
+    } else {
+        auto flds = jl_field_names(jt);
+        if (idx < jl_svec_len(flds)) {
+            auto name = jl_svec_ref(flds, idx);
+            assert(jl_is_symbol(name));
+            return jl_symbol_name((jl_sym_t*)name);
+        }
+    }
+    return "<unknown field>";
+}
+
 // If `nullcheck` is not NULL and a pointer NULL check is necessary
 // store the pointer to be checked in `*nullcheck` instead of checking it
 static jl_cgval_t emit_getfield_knownidx(jl_codectx_t &ctx, const jl_cgval_t &strct,
                                          unsigned idx, jl_datatype_t *jt,
                                          enum jl_memory_order order, Value **nullcheck)
 {
+    auto get_objname = [&]() {
+        return strct.V ? strct.V->getName() : StringRef("");
+    };
     jl_value_t *jfty = jl_field_type(jt, idx);
     bool isatomic = jl_field_isatomic(jt, idx);
     bool needlock = isatomic && !jl_field_isptr(jt, idx) && jl_datatype_size(jfty) > MAX_ATOMIC_SIZE;
     if (!isatomic && order != jl_memory_order_notatomic && order != jl_memory_order_unspecified) {
-        emit_atomic_error(ctx, "getfield: non-atomic field cannot be accessed atomically");
+        emit_atomic_error(ctx, ("getfield: non-atomic field " + StringRef(jl_symbol_name(jt->name->name)) + "." + get_fieldname(jt, idx) + " cannot be accessed atomically").str());
```

Comment:
Does this need to be matched in https://github.com/JuliaLang/julia/blob/c22b1c1393dccf3d2ded4d3ca30899955bb15264/src/builtins.c#L972?

Reply:
I see there are many more places where we'd have to update the error message to add field names, so I'll drop the improved error message change and follow this up in a separate PR.

Reply:
I also think we should stop trying to pre-generate these error strings as much, since there are potentially a lot of them, and they are not valuable to pre-compute (plus it makes it harder to keep them in-sync). An example of this alternative is 0718995f97c0d3df7f3f488523dd305c942efcf7 `jl_has_no_field_error`.